src/lib.rs

```diff
@@ -26,108 +26,115 @@
+//! on all connections.
 //!
-//! See the documentation of the `libp2p-core` crate for more details about upgrades.
-//!
-//! ## Topology
-//!
-//! The `Topology` trait is implemented for types that hold the layout of a network. When other
-//! components need the network layout to operate, they are passed an instance of a `Topology`.
+//! ## Network Behaviour
 //!
-//! The most basic implementation of `Topology` is the `MemoryTopology`, which is essentially a
-//! `HashMap`. Creating your own `Topology` makes it possible to add for example a reputation
-//! system.
+//! The [`NetworkBehaviour`] trait is implemented on types that provide some capability to the
+//! network. Examples of network behaviours include:
 //!
-//! ## Network behaviour
-//!
-//! The `NetworkBehaviour` trait is implemented on types that provide some capability to the
-//! network. Examples of network behaviours include: periodically ping the nodes we are connected
-//! to, periodically ask for information from the nodes we are connected to, connect to a DHT and
-//! make queries to it, propagate messages to the nodes we are connected to (pubsub), and so on.
+//!   * Periodically pinging other nodes on established connections.
+//!   * Periodically asking for information from other nodes.
+//!   * Querying information from a DHT and propagating it to other nodes.
 //!
 //! ## Swarm
 //!
-//! The `Swarm` struct contains all active and pending connections to remotes and manages the
-//! state of all the substreams that have been opened, and all the upgrades that were built upon
-//! these substreams.
-//!
-//! It combines a `Transport`, a `NetworkBehaviour` and a `Topology` together.
+//! A [`Swarm`] manages a pool of connections established through a [`Transport`]
+//! and drives a [`NetworkBehaviour`] through emitting events triggered by activity
+//! on the managed connections. Creating a [`Swarm`] thus involves combining a
+//! [`Transport`] with a [`NetworkBehaviour`].
 //!
-//! See the documentation of the `libp2p-core` crate for more details about creating a swarm.
+//! See the documentation of the [`core`] module for more details about swarms.
 //!
 //! # Using libp2p
 //!
-//! This section contains details about how to use libp2p in practice.
+//! The easiest way to get started with libp2p involves the following steps:
+//!
+//!   1. Creating a *base* implementation of [`Transport`] that is upgraded with
+//!      all the desired protocols, such as for transport security and multiplexing.
+//!   2. Creating a struct that implements the [`NetworkBehaviour`] trait and combines all the
+//!      desired network behaviours.
+//!   3. Instantiating a [`Swarm`] with the transport and network behaviour from the previous
+//!      steps.
 //!
-//! The most simple way to use libp2p consists in the following steps:
+//! The swarm instance can then be polled, e.g. with the [tokio] library, in order to
```

Comment:
Note that how we poll the swarm depends on the transport being used. If you use a `TcpConfig`, then you **must** use tokio, because the `TcpConfig` wraps around a `tokio_tcp::TcpStream`.

Reply:
Good point, I will try to make that clearer.

Reply:
Actually, I'm not sure I understand. Apart from being the obvious choice, why is it strictly necessary to use `tokio` to poll a swarm if the transport is based on a `tokio_tcp::TcpStream`?

Reply:
When you use a tokio runtime to poll, every time it polls a future tokio will define a thread-local-storage value containing a "reactor" (https://docs.rs/tokio-reactor). The reactor in this thread-local-storage is then picked up by the `tokio_tcp::TcpStream` to run itself. It is essentially an implicit parameter.

If you don't use a tokio runtime, then the `tokio_tcp::TcpStream` will produce an error because it can't find the reactor.

Reply:
Got it, thanks. I will update the documentation to make that clear then before merging.